bundles/org.eclipse.equinox.p2.publisher.eclipse/src/org/eclipse/equinox/p2/publisher/eclipse/BundlesAction.java

```diff
@@ -1151,4 +1170,51 @@ private void addPublishingErrorToFinalStatus(Throwable t, File bundleLocation) {
 		finalStatus.add(new Status(IStatus.ERROR, Activator.ID,
 				NLS.bind(Messages.exception_errorPublishingBundle, bundleLocation, t.getMessage()), t));
 	}
+
+	private void propagateProperties(Map<String, Object> source, Map<String, Object> target, String prefix) {
+		if (source == null) {
+			return;
+		}
+		for (Entry<String, Object> entry : source.entrySet()) {
+			String key = entry.getKey();
+			if (key.startsWith("x-")) { //$NON-NLS-1$
+				// not a standard directive / attribute if we want those we should only allow
+				// special know ones!
+				continue;
+			}
+			String prop = prefix + key;
+			Object value = entry.getValue();
+			if (value instanceof Object[] array) {
+				List<Object> list = List.of(array);
+				if (isValidValueType(list)) {
+					target.put(prop, list);
+				}
+			} else if (isValidValueType(prop)) {
+				target.put(prop, value);
+			}
+		}
+	}
+
+	private static boolean isValidValueType(Object prop) {
+		if (prop instanceof List<?>) {
+			int idx = 0;
+			for (Object scalar : (List<?>) prop) {
+				if (!isValidScalarType(scalar)) {
+					return false;
+				}
+			}
+			return true;
+		}
+		return isValidScalarType(prop);
+	}
+
+	private static boolean isValidScalarType(Object scalar) {
+		Class<?> clazz = scalar.getClass();
+		Stream<Class<?>> supportedClasses = Stream.of(Version.class, String.class, Long.class, Integer.class,
```

Comment:
Might it be better to create a static final constant constant for the list?

Reply:
Streams are usually not reusable since they have "terminal" operations closing them. You'd get a `java.lang.IllegalStateException: stream has already been operated upon or closed`.
Using Streams as storage is a bad idea; Collections are better for it. Directly using stream is better when processing the stream immediately, when there is no chance of it being referenced anywhere else (because it's not accessible or because same expression would lead to a new stream instance); but here it's not the case.